api/http_client.go

```diff
@@ -115,6 +122,192 @@ func ExtractHeader(name string, dest *string) func(http.RoundTripper) http.Round
+// all non-printable characters between \u0000 and \u00FF so that the terminal will
+// not interpret them.
+func SanitizeASCIIControlCharacters(rt http.RoundTripper) http.RoundTripper {
+	return &funcTripper{roundTrip: func(req *http.Request) (*http.Response, error) {
+		res, err := rt.RoundTrip(req)
+		if err != nil ||
+			!jsonTypeRE.MatchString(res.Header.Get("Content-Type")) ||
+			res.ContentLength == 0 {
+			return res, err
+		}
+		var sanitized bytes.Buffer
+		err = replaceControlCharacters(res.Body, &sanitized)
+		if err != nil {
+			err = fmt.Errorf("ascii control characters sanitization error: %w", err)
+		}
+		res.Body.Close()
+		res.Body = io.NopCloser(&sanitized)
+		res.ContentLength = int64(sanitized.Len())
+		return res, err
+	}}
+}
+
+// replaceControlCharacters is a sliding window alogorithm that
+// detects C0 and C1 ASCII control sequences as they are read
+// from r and replaces them with equivelent inert characters and
+// writes them to w. Characters that are not part of a control
+// sequence are written as is to w.
+func replaceControlCharacters(r io.Reader, w io.Writer) error {
+	// Byte representation of the string sequence "\u00" which is the prefix of
+	// all C0 and C1 control sequeneces.
+	find := []byte{92, 117, 48, 48}
+	// Length of find sequence.
+	size := 4
+	// Create an index variable to match bytes.
+	idx := 0
+	// Used for reading single byte.
+	b := make([]byte, 1)
+	// Used for reading two bytes.
+	c := make([]byte, 2)
+	// Used to keep track of escape characters.
+	var addBackslash bool
+	// Used to keep track of when to read.
+	var skipRead bool
+
+	for {
+		if !skipRead {
+			if n, err := r.Read(b); err != nil {
+				if !errors.Is(err, io.EOF) {
+					return err
+				}
+				if n > 0 {
+					if _, err = w.Write(b); err != nil {
+						return err
+					}
+				}
+				break
+			}
+		}
+		skipRead = false
+
+		// Match.
+		if b[0] == find[idx] {
+			idx++
+			if idx == size {
+				if _, err := r.Read(c); err != nil {
+					return err
+				}
+				s := append(find, c[0], c[1])
+				repl, found := mapControlCharacterToCaret(s)
+				if found && addBackslash {
+					repl = append([]byte{92}, repl...)
+				}
+				if _, err := w.Write(repl); err != nil {
+					return err
+				}
+				idx = 0
+				addBackslash = false
+			}
+			continue
+		}
+
+		// No match but with previous match.
+		if idx != 0 {
+			if _, err := w.Write(find[:idx]); err != nil {
+				return err
+			}
+			if idx == 1 {
+				addBackslash = !addBackslash
+			}
+			idx = 0
+			skipRead = true
+			continue
+		}
+
+		// No match and no previous match.
+		if _, err := w.Write(b); err != nil {
+			return err
+		}
+		idx = 0
+		addBackslash = false
+	}
+
+	return nil
+}
+
+// mapControlCharacterToCaret maps C0 control sequences to caret notation and
+// C1 control sequences to hex notation. C1 control sequences do
+// not have caret notation representation.
+func mapControlCharacterToCaret(b []byte) ([]byte, bool) {
```

Comment:
Love the bit of added UX by mapping unsafe escape codes to their printable equivalents 👍